### PR TITLE
Implement OOP matrix cache

### PR DESCRIPTION
**Pull request by zfergus**
This allows you to build with either the sparse matrix cache or a dense matrix. I am not sure if it is needed anymore.

### CODE REVIEW

**Comment by teseoch** [comment]
should we merge this?

File: src/polyfem/assembler/Assembler.cpp [hunk 1]
```diff
@@ -18,35 +18,54 @@ namespace polyfem::assembler
...
+				return *this;
+			}
+
 			void init(const int buffer_size, const int rows, const int cols)
 			{
 				// assert(rows == cols);
-				cache.reserve(buffer_size);
-				cache.init(rows, cols);
+				// TODO: Expose this choice
+				if (rows <= 1000 || cols <= 1000)
```
Comment (zfergus) [inline src/polyfem/assembler/Assembler.cpp:55]: Revert this for now

File: src/polyfem/solver/forms/ElasticForm.cpp [hunk 2]
```diff
@@ -12,6 +12,8 @@ using namespace polyfem::utils;
 
 namespace polyfem::solver
 {
+	static constexpr size_t MAX_DENSE_CACHE_SIZE = 300;
```
Comment (zfergus) [inline src/polyfem/solver/forms/ElasticForm.cpp:15]: Revert this for now

File: src/polyfem/solver/forms/ElasticForm.cpp [hunk 3]
```diff
@@ -48,6 +50,10 @@ namespace polyfem::solver
 	{
 		if (assembler_.is_linear())
 			compute_cached_stiffness();
+		if (n_bases * (is_volume ? 3 : 2) <= MAX_DENSE_CACHE_SIZE) // TODO: Expose this choice
```
Comment (zfergus) [inline src/polyfem/solver/forms/ElasticForm.cpp:53]: Revert this for now

File: src/polyfem/solver/forms/ElasticForm.hpp [hunk 4]
```diff
@@ -77,8 +77,8 @@ namespace polyfem::solver
 		const double dt_;
 		const bool is_volume_;
 
-		StiffnessMatrix cached_stiffness_;           ///< Cached stiffness matrix for linear elasticity
-		mutable utils::SparseMatrixCache mat_cache_; ///< Matrix cache (mutable because it is modified in second_derivative_unweighted)
+		StiffnessMatrix cached_stiffness_;                      ///< Cached stiffness matrix for linear elasticity
+		mutable std::shared_ptr<utils::MatrixCache> mat_cache_; ///< Matrix cache (mutable because it is modified in second_derivative_unweighted)
```
Comment (zfergus) [inline src/polyfem/solver/forms/ElasticForm.hpp:81]: unique pointer?

**Comment by zfergus** [comment]
> should we merge this?

Yes, once the CI passes.

**Pull request merged**